### PR TITLE
Use HTML5 download parameter for download CTAs

### PR DESCRIPTION
**Pull request by gidsg**
Previously we were forcing download content disposition in all cases. Now we only set it where there is a specific download call to action.

Notes
* This relies on the [HTML5 `download` attribute](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a#attr-download) which has [limited browser support](https://caniuse.com/#feat=download). 
* I am no longer setting the custom filenames which were triggered on download previously. So we rely on the default filename from the server e.g. `records.json` when the file is downloaded.

### CODE REVIEW

File: src/main/java/uk/gov/register/resources/EntryResource.java [hunk 1]
```diff
@@ -59,8 +59,7 @@ public EntryResource(RegisterReadOnly register, ViewFactory viewFactory, Request
     @Timed
     public Optional<EntryListView> findByEntryNumber(@PathParam("entry-number") int entryNumber) {
         Optional<Entry> entry = register.getEntry(entryNumber);
-        headerProvider.setAttachmentContentDisposition(String.valueOf(entryNumber));
-        return entry.map(function -> new EntryListView(Collections.singletonList(function)));
+        return entry.isPresent() ? Optional.of(new EntryListView(Arrays.asList(entry.get()))) : Optional.empty();
```
Comment (fjavierm) [inline src/main/java/uk/gov/register/resources/EntryResource.java]: Why this conditional structure? The map function does exactly this.

File: src/main/resources/templates/fragments/download-and-preview.html [hunk 2]
```diff
@@ -5,16 +5,16 @@
...
+         data-download-type="spreadsheet" download="">spreadsheet</a>.
       <th:block th:if="${pagination.totalEntries} le 100">You can also preview as</th:block>
       <th:block th:if="${pagination.totalEntries} gt 100">You can also preview a sample as</th:block>
       <a th:href="${'/preview-' + records_or_entries_string + '/json'}">JSON</a>,
       <a th:href="${'/preview-' + records_or_entries_string + '/yaml'}">YAML</a> or
       <a th:href="${'/preview-' + records_or_entries_string + '/ttl'}">TTL</a>.
     </p>
   </div>
-</html>
+</html>
```
Comment (fjavierm) [inline src/main/resources/templates/fragments/download-and-preview.html]: NIT: Not line at the end of the file

File: src/main/resources/templates/fragments/download-and-preview-home.html [hunk 3]
```diff
@@ -4,9 +4,9 @@
     <p class="data-representations">
       <th:block th:if="${totalRecords} le 5000">Download a copy of this register as</th:block>
       <th:block th:if="${totalRecords} gt 5000">Download the first 5,000 records as</th:block>
-      <a href="/records.csv?page-size=5000" rel="alternate" class="js-download" data-download-type="CSV">CSV</a>,
-      <a href="/records.tsv?page-size=5000" rel="alternate" class="js-download" data-download-type="TSV">TSV</a> or as a
-      <a href="/records.xlsx?page-size=5000" rel="alternate" class="js-download" data-download-type="XLSX">spreadsheet</a>.
+      <a href="/records.csv?page-size=5000" rel="alternate" class="js-download" data-download-type="CSV" download="">CSV</a>,
```
Comment (fjavierm) [inline src/main/resources/templates/fragments/download-and-preview-home.html]: @danmitchell- can correct me on this but as far as I know, Custom Data Attributes start with `data-` in HTML5.

File: src/main/java/uk/gov/register/resources/RecordResource.java [hunk 4]
```diff
@@ -20,18 +20,18 @@
 @Path("/")
 public class RecordResource {
     private final HttpServletResponseAdapter httpServletResponseAdapter;
+    private final RequestContext requestContext;
```
Comment (fjavierm) [inline src/main/java/uk/gov/register/resources/RecordResource.java]: `requestContext` is only used in the constructor, why are you adding it as property?

File: src/main/java/uk/gov/register/resources/RecordResource.java [hunk 5]
```diff
@@ -20,18 +20,18 @@
...
     private final RegisterReadOnly register;
     private final ViewFactory viewFactory;
     private final RegisterName registerPrimaryKey;
```
Comment (fjavierm) [inline src/main/java/uk/gov/register/resources/RecordResource.java]: `registerPrimaryKey` is not used any more, except in the constructor, can it be removed?

File: src/main/resources/templates/fragments/download-and-preview-home.html (hunk 3 above)
Comment (gidsg) [reply src/main/resources/templates/fragments/download-and-preview-home.html]: @danmitchell- this is not a data attribute. See https://www.w3schools.com/tags/att_a_download.asp

Comment (fjavierm) [reply src/main/resources/templates/fragments/download-and-preview-home.html]: @gidsg I saw that, but I thought it was just `download` without `=""` . I was not sure what it was. Just ignore the comment then

Comment (danmitchell-) [reply src/main/resources/templates/fragments/download-and-preview-home.html]: Correct. Its not a `data` attribute.

**Comment by gidsg** [comment]
@fjavierm I think once you are doing an `Open in New Tab` generally the browser is treating that as a new request so ignores the `download` attribute. I am not sure how big an issue this is.

**Comment by fjavierm** [comment]
@gidsg My point is that with the `ContentDisposition` attachement, this was working as expected.

**Comment by gidsg** [comment]
@fjavierm I am not sure if the user selects Open In New Tab what exactly they are expecting. If this is really critical we could implement it server-side as a query string parameter which sets the `ContentDisposition` but I would argue the intended behaviour is ambiguous in this case. Happy to talk with you about this.

File: src/main/resources/templates/fragments/download-and-preview-home.html (hunk 3 above)
Comment (danmitchell-) [reply src/main/resources/templates/fragments/download-and-preview-home.html]: Not sure if im missing something here but you should be passing the URL into the download attribute i.e. `<a href="/preview-records/json" download="/preview-records/json">JSON</a>`

Comment (gidsg) [reply src/main/resources/templates/fragments/download-and-preview-home.html]: No, it uses the URL from the href. The value is optional and is used to override the filename given by the server.

Comment (danmitchell-) [reply src/main/resources/templates/fragments/download-and-preview-home.html]: Ah yes. Not sure how much I trust that but 👌

**Pull request merged**